community/kernel/src/main/java/org/neo4j/kernel/impl/proc/OutputMappers.java

```diff
@@ -0,0 +1,186 @@
+import static java.util.Arrays.asList;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Takes user-defined record classes, and does two things: Describe the class as a {@link ProcedureSignature}, and provide a mechanism to convert
+ * an instance of the class to Neo4j-typed Object[].
+ */
+public class OutputMappers
+{
+    public OutputMappers( TypeMappers typeMappers )
+    {
+        this.typeMappers = typeMappers;
+    }
+
+    /**
+     * A compiled mapper, takes an instance of a java class, and converts it to an Object[] matching the specified {@link #signature()}.
+     */
+    public static class OutputMapper
+    {
+        private final List<FieldSignature> signature;
+        private final FieldMapper[] fieldMappers;
+
+        public OutputMapper( FieldSignature[] signature, FieldMapper[] fieldMappers )
+        {
+            this.signature = asList( signature );
+            this.fieldMappers = fieldMappers;
+        }
+
+        public List<FieldSignature> signature()
+        {
+            return signature;
+        }
+
+        public Object[] apply( Object record ) throws ProcedureException
+        {
+            Object[] output = new Object[fieldMappers.length];
+            for ( int i = 0; i < fieldMappers.length; i++ )
+            {
+                output[i] = fieldMappers[i].apply( record );
+            }
+            return output;
+        }
+    }
+
+    /**
+     * Extracts field value from an instance and converts it to a Neo4j typed value.
+     */
+    private static class FieldMapper
+    {
+        private final MethodHandle getter;
+        private final TypeMappers.NeoValueConverter mapper;
+
+        public FieldMapper( MethodHandle getter, TypeMappers.NeoValueConverter mapper )
+        {
+            this.getter = getter;
+            this.mapper = mapper;
+        }
+
+        Object apply( Object record ) throws ProcedureException
+        {
+            Object invoke = getValue( record );
+            return mapper.toNeoValue( invoke );
+        }
+
+        private Object getValue( Object record ) throws ProcedureException
+        {
+            try
+            {
+                return getter.invoke( record );
+            }
+            catch ( Throwable throwable )
+            {
+                throw new ProcedureException( Status.Procedure.CallFailed, throwable, "Unable to read value from record: %s", throwable.getMessage() );
+            }
+        }
+    }
+
+    private final Lookup lookup = MethodHandles.lookup();
+    private final TypeMappers typeMappers;
+
+    /**
+     * Build an output mapper for the return type of a given method.
+     *
+     * @param method the procedure method
+     * @return an outputmapper for the return type of the method.
+     * @throws ProcedureException
+     */
+    public OutputMapper mapper( Method method ) throws ProcedureException
+    {
+        Class<?> cls = method.getReturnType();
+        if ( cls != Stream.class )
```

Comment:
If i was following the code correctly this check is no longer valid since now its RawIterator right?